### PR TITLE
[exporter/skywalking] Add skywalking metrics exporter

### PR DESCRIPTION
**Pull request by liqiangz**
**Description:**
 Add skywalking metrics exporter

**Link to tracking Issue:** 
apache/skywalking#6210

**Testing:** <Describe what testing was performed and which tests were added.>
metricrecord_to_metricdata_test.go

**Documentation:** <Describe the documentation added.>

### CODE REVIEW

**Comment by mx-psi** [comment]
@liqiangz can you fix the conflicts?

**Comment by liqiangz** [comment]
> @liqiangz can you fix the conflicts?

@mx-psi fixed

File: exporter/skywalkingexporter/logrecord_to_logdata.go [hunk 1]
```diff
@@ -72,9 +72,6 @@ func resourceToLogData(resource pdata.Resource, logData *logpb.LogData) {
 	}
 
 	attrs.Range(func(k string, v pdata.AttributeValue) bool {
-		if k == conventions.AttributeServiceName || k == conventions.AttributeServiceInstanceID {
-			return true
-		}
```
Comment (mx-psi) [inline exporter/skywalkingexporter/logrecord_to_logdata.go:77]: not sure what the context of this is?

File: exporter/skywalkingexporter/metricrecord_to_metricdata.go [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+		sv.SingleValue.Name = name
+		switch dataPoint.Type() {
+		case pdata.MetricValueTypeInt:
+			sv.SingleValue.Value = float64(dataPoint.IntVal())
+		case pdata.MetricValueTypeDouble:
+			sv.SingleValue.Value = dataPoint.DoubleVal()
+		}
+		meterData.Metric = sv
+		meterData.Service = defaultServiceName
+		meterData.ServiceInstance = defaultServiceInstance
```
Comment (mx-psi) [inline exporter/skywalkingexporter/metricrecord_to_metricdata.go:69]: You may be interested in https://github.com/open-telemetry/opentelemetry-specification/blob/da74730bf835010229a9612c281f052e26553218/specification/resource/semantic_conventions/README.md#service (just a FYI, I think this can be a separate PR)

File: exporter/skywalkingexporter/metricrecord_to_metricdata.go [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+		svc.SingleValue.Value = float64(dataPoint.Count())
+		meterDataCount.Metric = svc
+		meterDataCount.Service = defaultServiceName
+		meterDataCount.ServiceInstance = defaultServiceInstance
+		metrics = append(metrics, meterDataCount)
+	}
+	return metrics
+}
+
+func metricDataToMetricData(md pdata.Metric, defaultLabels []*metricpb.Label) (metrics []*metricpb.MeterData) {
```
Comment (mx-psi) [inline exporter/skywalkingexporter/metricrecord_to_metricdata.go:193]: What about `MetricDataTypeExponentialHistogram`? It's not necessary to do it on this PR, and it's not a widely used type yet, but you may want to create an issue to add support for this type. 

Also, it's possible that OTLP will have new types over time. It may be a good idea to have some observability feature for the user to know when datapoints are dropped because a metric type is not supported by this exporter.

File: exporter/skywalkingexporter/metricrecord_to_metricdata.go (hunk 2 above)
Comment (liqiangz) [reply exporter/skywalkingexporter/metricrecord_to_metricdata.go]: Updated

File: exporter/skywalkingexporter/logrecord_to_logdata.go (hunk 1 above)
Comment (liqiangz) [reply exporter/skywalkingexporter/logrecord_to_logdata.go:77]: These two fields were excluded because they were already in service and serviceInstance. But sometimes when using tag to analyze data on the skywalking side, this data is still needed.

File: exporter/skywalkingexporter/metricrecord_to_metricdata.go (hunk 3 above)
Comment (liqiangz) [reply exporter/skywalkingexporter/metricrecord_to_metricdata.go]: Do you have any suggestions about observability.  Maybe Indicate the supported types in the document, or print the warn log when encountering types that cannot be processed. I see that the existing exporters basically ignore this.

Comment (mx-psi) [reply exporter/skywalkingexporter/metricrecord_to_metricdata.go]: >  Maybe Indicate the supported types in the document, or print the warn log when encountering types that cannot be processed

Both seem like good things to have (maybe with a debug level log instead, to be non-spammy by default?). It's true that other exporters ignore this issue, but I think it's important to have *something* until you support `ExponentialHistogram`, so that end-users have some lead when understanding why they don't see the metrics in your backend.

**Review by mx-psi — APPROVED** [review]
@liqiangz Since the missing features can be added in a future PR and everything else LGTM, I am approving this in the interest of getting it merged before the holidays. Will wait until tomorrow CEST morning to mark as `ready-to-merge` in case you think otherwise.

**Comment by mx-psi** [comment]
@liqiangz please fix the merge conflict

File: exporter/skywalkingexporter/go.mod [hunk 4]
```diff
@@ -9,7 +9,14 @@ require (
 )
 
 require (
-	cloud.google.com/go v0.81.0 // indirect
+	github.com/open-telemetry/opentelemetry-collector-contrib/internal/coreinternal v0.41.0
```
Comment (jpkrohling) [inline exporter/skywalkingexporter/go.mod:12]: Please, cleanup your go.mod a bit: leave only two sections, with the first being the direct dependencies and the second the indirect ones.

File: exporter/skywalkingexporter/metricrecord_to_metricdata.go [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+	if serviceInstanceID, ok := attrs.Get(conventions.AttributeServiceInstanceID); ok {
+		serviceInstance = serviceInstanceID.AsString()
+	} else {
+		serviceInstance = defaultServiceInstance
+	}
+	return service, serviceInstance
+}
+
+func numberMetricsToData(name string, data pdata.NumberDataPointSlice, defaultLabels []*metricpb.Label) (metrics []*metricpb.MeterData) {
+	metrics = make([]*metricpb.MeterData, 0)
```
Comment (jpkrohling) [inline exporter/skywalkingexporter/metricrecord_to_metricdata.go:59]: Will the len of the resulting `metrics` always be the same as `data.Len()`? If so, you might want to initialize this with the final size, potentially saving some resize operations.

File: exporter/skywalkingexporter/metricrecord_to_metricdata.go [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+			sv.SingleValue.Value = dataPoint.DoubleVal()
+		}
+		meterData.Metric = sv
+		metrics = append(metrics, meterData)
+	}
+	return metrics
+}
+
+func doubleHistogramMetricsToData(name string, data pdata.HistogramDataPointSlice, defaultLabels []*metricpb.Label) (metrics []*metricpb.MeterData) {
+	metrics = make([]*metricpb.MeterData, 0)
```
Comment (jpkrohling) [inline exporter/skywalkingexporter/metricrecord_to_metricdata.go:90]: Same here.

File: exporter/skywalkingexporter/metricrecord_to_metricdata.go [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+		svc.SingleValue.Name = name + "_count"
+		svc.SingleValue.Value = float64(dataPoint.Count())
+		meterDataCount.Metric = svc
+		metrics = append(metrics, meterDataCount)
+	}
+	return metrics
+}
+
+func doubleSummaryMetricsToData(name string, data pdata.SummaryDataPointSlice, defaultLabels []*metricpb.Label) (metrics []*metricpb.MeterData) {
+	metrics = make([]*metricpb.MeterData, 0)
```
Comment (jpkrohling) [inline exporter/skywalkingexporter/metricrecord_to_metricdata.go:144]: And here (and potentially elsewhere)

File: exporter/skywalkingexporter/skywalking_benchmark_test.go [hunk 8]
```diff
@@ -145,7 +145,7 @@ func doInit(numStream int, t *testing.T) (*swExporter, *grpc.Server, *mockLogHan
 		},
 	}
 
-	oce := newExporter(context.Background(), tt, componenttest.NewNopTelemetrySettings())
+	oce := newLogsExporter(context.Background(), tt, componenttest.NewNopTelemetrySettings())
```
Comment (jpkrohling) [inline exporter/skywalkingexporter/skywalking_benchmark_test.go:148]: Please ensure that tech debt is addressed before adding new features: https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/6116

File: exporter/skywalkingexporter/metricrecord_to_metricdata.go [hunk 9]
```diff
@@ -0,0 +1,241 @@
...
+	}
+	return service, serviceInstance
+}
+
+func numberMetricsToData(name string, data pdata.NumberDataPointSlice, defaultLabels []*metricpb.Label) (metrics []*metricpb.MeterData) {
+	metrics = make([]*metricpb.MeterData, 0, data.Len())
+	for i := 0; i < data.Len(); i++ {
+		dataPoint := data.At(i)
+		attributeMap := dataPoint.Attributes()
+		labels := make([]*metricpb.Label, 0)
```
Comment (jpkrohling) [inline exporter/skywalkingexporter/metricrecord_to_metricdata.go:63]: This also seems to have a fixed size (data point attributes' length plus default labels' length). If you prefer to address this in a follow-up PR, create an issue to track it.

This applies to other places in this PR.

**Review by jpkrohling — APPROVED** [review]

File: exporter/skywalkingexporter/skywalking_benchmark_test.go (hunk 8 above)
Comment (liqiangz) [reply exporter/skywalkingexporter/skywalking_benchmark_test.go:148]: I will fix it in another pr.

File: exporter/skywalkingexporter/metricrecord_to_metricdata.go (hunk 9 above)
Comment (liqiangz) [reply exporter/skywalkingexporter/metricrecord_to_metricdata.go]: fixed.

**Review by jpkrohling — APPROVED** [review]
